drivers/entropy/entropy_nrf5.c

```diff
@@ -63,204 +63,165 @@
+		 (CONFIG_ENTROPY_NRF5_THR_POOL_SIZE - 1)) == 0,
+		 "The CONFIG_ENTROPY_NRF5_THR_POOL_SIZE must be a power of 2!");
 
 struct entropy_nrf5_dev_data {
 	struct k_sem sem_lock;
 	struct k_sem sem_sync;
 
-	RAND_DEFINE(thr, RAND_THREAD_LEN);
-	RAND_DEFINE(isr, RAND_ISR_LEN);
+	RNG_POOL_DEFINE(isr, CONFIG_ENTROPY_NRF5_ISR_POOL_SIZE);
+	RNG_POOL_DEFINE(thr, CONFIG_ENTROPY_NRF5_THR_POOL_SIZE);
 };
 
 #define DEV_DATA(dev) \
 	((struct entropy_nrf5_dev_data *)(dev)->driver_data)
 
-#pragma GCC push_options
-#if defined(CONFIG_BT_CTLR_FAST_ENC)
-#pragma GCC optimize ("Ofast")
-#endif
-static inline u8_t get(struct rand *rng, u8_t octets, u8_t *rand)
+static int random_byte_get(void)
 {
-	u8_t first, last, avail, remaining, *d, *s;
-
-	__ASSERT_NO_MSG(rng);
-
-	first = rng->first;
-	last = rng->last;
-
-	d = &rand[octets];
-	s = &rng->rand[first];
-
-	if (first <= last) {
-		/* copy octets from contiguous memory */
-		avail = last - first;
-		if (octets < avail) {
-			remaining = avail - octets;
-			avail = octets;
-		} else {
-			remaining = 0;
-		}
-
-		first += avail;
-		octets -= avail;
-
-		while (avail--) {
-			*(--d) = *s++;
-		}
-
-		rng->first = first;
-	} else {
-		/* copy octets from split halves - until end of array */
-		avail = rng->count - first;
-		if (octets < avail) {
-			remaining = avail + last - octets;
-			avail = octets;
-			first += avail;
-		} else {
-			remaining = last;
-			first = 0;
-		}
+	int retval = -EAGAIN;
+	unsigned int key;
 
-		octets -= avail;
+	key = irq_lock();
 
-		while (avail--) {
-			*(--d) = *s++;
-		}
+	if (nrf_rng_event_get(NRF_RNG_EVENT_VALRDY)) {
+		retval = nrf_rng_random_value_get();
+		nrf_rng_event_clear(NRF_RNG_EVENT_VALRDY);
+	}
 
-		/* copy from beginning of array - until ring buffer last idx */
-		if (octets && last) {
-			s = &rng->rand[0];
+	irq_unlock(key);
 
-			if (octets < last) {
-				remaining = last - octets;
-				last = octets;
-			} else {
-				remaining = 0;
-			}
-
-			first = last;
-			octets -= last;
+	return retval;
+}
 
-			while (last--) {
-				*(--d) = *s++;
-			}
-		}
+#pragma GCC push_options
+#if defined(CONFIG_BT_CTLR_FAST_ENC)
+#pragma GCC optimize ("Ofast")
+#endif
+static u16_t rng_pool_get(struct rng_pool *rngp, u8_t *buf, u16_t len)
+{
+	u32_t last  = rngp->last;
+	u32_t mask  = rngp->mask;
+	u8_t *dst   = buf;
+	u32_t first, available;
+	u32_t other_read_in_progress;
+	unsigned int key;
+
+	key = irq_lock();
+	first = rngp->first_alloc;
+
+	other_read_in_progress = (rngp->first_read ^ first);
+	available = (last - first) & mask;
+	if (available < len)
+		len = available;
+
+	rngp->first_alloc = (first + len) & mask;
+	irq_unlock(key);
+
+	while (likely(len--)) {
+		*dst++ = rngp->buffer[first];
+		first = (first + 1) & mask;
+	}
 
-		rng->first = first;
+	if (likely(!other_read_in_progress)) {
+		key = irq_lock();
+		rngp->first_read = rngp->first_alloc;
+		irq_unlock(key);
 	}
 
-	if (remaining < rng->threshold) {
+	len = dst - buf;
+	available = available - len;
```

Comment:
You could move this line before `while (likely(len--))`. Then this reconstruction of `len` will be not needed.
Actually, a smart compiler should be able to make such optimization, but...

Reply:
Right. I forgot about the return value.